### PR TITLE
Add support for Slack's "Blocks" and OAuth instead of webhooks

### PR DESCRIPTION
**Pull request by mikeymclellan**

### CODE REVIEW

File: src/Client.php [hunk 1]
```diff
@@ -68,20 +76,42 @@ public function __construct(string $endpoint, array $options = [], HttpClient $h
...
-                    ),
-                ]
+                $plugins
             ),
             MessageFactoryDiscovery::find()
         );
+        return $this;
+    }
+
+    public function setOptions(array $options, string $endpoint = self::SLACK_POST_MESSAGE_URL, HttpClient $httpClient = null): self
```
Comment (hashnz) [inline src/Client.php:110]: won't this override the constructor $endpoint if you just call it with $options?

Comment (mikeymclellan) [reply src/Client.php:110]: Yep, which is a bit lame. It's a bit of a balancing act trying to not break backward-compatibility. 

Slack has deprecated webhook endpoints so ideally `endpoint` would just be an option with a sensible default. But unfortunately, the constructor has it as the first parameter. 

The two ways I see this being called are the old web-hook way:

```
$client = new Nexy\Slack\Client('https://hooks.slack.com/...', $options);
```

And the new oauth way:

```
$client = (new Nexy\Slack\Client())->setOptions($options);
```

I guess I could get rid of the `setOptions` function and then people would have to instantiate like this:

```
$client = new Nexy\Slack\Client(Nexy\Slack\Client::SLACK_POST_MESSAGE_URL, $options);
```

But I kinda thought that was a little yuck.

🤔

Comment (hashnz) [reply src/Client.php:110]: ah I see, could you just call setEndpoint from the constructor n remove it from setOptions? up to you :+1:

**Review by hashnz — APPROVED** [review]

File: src/Client.php (hunk 1 above)
Comment (mikeymclellan) [reply src/Client.php:110]: Yeah, I wanted to do that too. But unfortunately one of the options; `oauth_token`, requires me to update `$this->httpClient` which is what `setEndpoint()` does. 

So I could optimise `setOptions` so it modifies the existing `httpClient` I guess. But I started to think I'd spent enough time on this library. And it probably would be better to break backward-compatibility at that point. 

If nexylan/slack doesn't approve this PR I'll refactor and tidy this up.

Comment (hashnz) [reply src/Client.php:110]: :100:

**Pull request merged**